lib/Core/Search/Legacy/Query/Content/CriterionHandler/LocationQuery.php

```diff
@@ -45,17 +45,17 @@ public function handle(
     {
         /** @var \eZ\Publish\API\Repository\Values\Content\Query\Criterion $filter */
         $filter = $criterion->value;
-        $subSelect = $this->connection->createQueryBuilder();
+        $subSelect = new SubQueryQueryBuilder($this->connection, $queryBuilder);
```

Comment:
Not really sure why this is needed? Can't you just use `$queryBuilder` directly, since all `SubQueryQueryBuilder` does is forward the `createNamedParameter` to it?

Reply:
You cannot pass the `$queryBuilder` to `$this->locationCriteriaConverter->convertCriteria` as it needs to add to the `$subSelect` and not the outer `$queryBuilder`

`SubQueryQueryBuilder` is needed so the added `$subSelect->getSQL()`  does not have parameters with name clashes of the outer.  Perhaps a better name could be `SubSelectQueryBuilder`

For example:

$queryBuilder->createNamedParameter would generate a name of `:dcValue1`
$subSelect->createNamedParameter  would then generate a name of `:dcValue1`

The parameter names clash so the `SubQueryQueryBuilder` uses the outer `$queryBuilder->createNamedParameter`  internally so you get $subSelect->createNamedParameter  would then generate a name of `:dcValue2` no clash.

Hope that helps clear it up.

Reply:
It does, yes!

Just a thought: can the `$queryBuilder` object be cloned and existing parameters be cleared (or even instantiate a new query builder dierectly), so we don't need to maintain an additional class?

If not,  I guess this is okay as is.

Reply:
I tried cloning and it did not work.  Also even if cloned you cannot guarantee that no name clashes would take place.

Would have been nice if the queryBuilder always used unique generated names but if you look at how it was written not possible.  I think we are stuck with this.